rand.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2018 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package saltpack
+
+import (
+	cryptorand "crypto/rand"
+	"encoding/binary"
+	"io"
+)
+
+// csprngReadFull is a thin wrapper around io.ReadFull on a given
+// CSPRNG that also (paranoidly) checks the length.
+func csprngReadFull(csprng io.Reader, b []byte) error {
+	n, err := io.ReadFull(csprng, b)
+	if err != nil {
+		return err
+	}
+	if n != len(b) {
+		return ErrInsufficientRandomness
+	}
+	return nil
+}
+
+// csprngRead is like crypto/rand.Read, except it uses csprngReadFull
+// instead of io.ReadFull.
+func csprngRead(b []byte) error {
+	return csprngReadFull(cryptorand.Reader, b)
+}
+
+// csprngUint32, given a CSPRNG, returns a uniformly distributed
+// random number in [0, 2³²).
+func csprngUint32(csprng io.Reader) (uint32, error) {
+	var buf [4]byte
+	err := csprngReadFull(csprng, buf[:])
+	if err != nil {
+		return 0, err
+	}
+
+	return binary.BigEndian.Uint32(buf[:]), nil
+}
+
+// csprngUint32n, given a CSPRNG, returns, as a uint32, a uniformly
+// distributed random number in [0, n). It is adapted from
+// math/rand.int31n from go 1.10.
+//
+// For implementation details, see:
+// https://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction
+// https://lemire.me/blog/2016/06/30/fast-random-shuffling
+func csprngUint32n(csprng io.Reader, n uint32) (uint32, error) {
+	v, err := csprngUint32(csprng)
+	if err != nil {
+		return 0, err
+	}
+	prod := uint64(v) * uint64(n)
+	low := uint32(prod)
+	if low < n {
+		thresh := -n % n
```

Comment:
In the original implementation, `n` is an `int32`. Here, it's a `uint32`. Is negating it still kosher?

Reply:
in the original interpretation `n` is an `int32`, but they do `uint32(-n) % uint32(n)`. I checked and they do mean `(2**32 - n) % n`.